validate/validate.go

```diff
@@ -139,78 +148,83 @@ func (v *Validator) CheckRoot() (msgs []string) {
 	}
 
 	rootParent := filepath.Dir(absRootPath)
 	if absRootPath == string(filepath.Separator) || rootParent != absBundlePath {
-		msgs = append(msgs, fmt.Sprintf("root.path is %q, but it MUST be a child of %q", v.spec.Root.Path, absBundlePath))
+		errs = multierror.Append(errs,
+			NewError(ArtifactsInSingleDir, fmt.Sprintf("root.path is %q, but it MUST be a child of %q", v.spec.Root.Path, absBundlePath), rspec.Version))
 	}
 
 	if v.platform == "windows" {
 		if v.spec.Root.Readonly {
-			msgs = append(msgs, "root.readonly field MUST be omitted or false when target platform is windows")
+			errs = multierror.Append(errs,
+				NewError(ReadonlyOnWindows, "root.readonly field MUST be omitted or false when target platform is windows", rspec.Version))
 		}
 	}
 
 	return
 }
 
 // CheckSemVer checks v.spec.Version
-func (v *Validator) CheckSemVer() (msgs []string) {
+func (v *Validator) CheckSemVer() (errs error) {
 	logrus.Debugf("check semver")
 
 	version := v.spec.Version
 	_, err := semver.Parse(version)
 	if err != nil {
-		msgs = append(msgs, fmt.Sprintf("%q is not valid SemVer: %s", version, err.Error()))
+		errs = multierror.Append(errs,
+			NewError(SpecVersion, fmt.Sprintf("%q is not valid SemVer: %s", version, err.Error()), rspec.Version))
 	}
 	if version != rspec.Version {
-		msgs = append(msgs, fmt.Sprintf("internal error: validate currently only handles version %s, but the supplied configuration targets %s", rspec.Version, version))
+		errs = multierror.Append(errs, fmt.Errorf("internal error: validate currently only handles version %s, but the supplied configuration targets %s", rspec.Version, version))
 	}
 
 	return
 }
 
 // CheckHooks check v.spec.Hooks
-func (v *Validator) CheckHooks() (msgs []string) {
+func (v *Validator) CheckHooks() (errs error) {
 	logrus.Debugf("check hooks")
 
 	if v.spec.Hooks != nil {
-		msgs = append(msgs, checkEventHooks("pre-start", v.spec.Hooks.Prestart, v.HostSpecific)...)
-		msgs = append(msgs, checkEventHooks("post-start", v.spec.Hooks.Poststart, v.HostSpecific)...)
-		msgs = append(msgs, checkEventHooks("post-stop", v.spec.Hooks.Poststop, v.HostSpecific)...)
+		errs = multierror.Append(errs, checkEventHooks("pre-start", v.spec.Hooks.Prestart, v.HostSpecific))
+		errs = multierror.Append(errs, checkEventHooks("post-start", v.spec.Hooks.Poststart, v.HostSpecific))
+		errs = multierror.Append(errs, checkEventHooks("post-stop", v.spec.Hooks.Poststop, v.HostSpecific))
 	}
 
 	return
 }
 
-func checkEventHooks(hookType string, hooks []rspec.Hook, hostSpecific bool) (msgs []string) {
+func checkEventHooks(hookType string, hooks []rspec.Hook, hostSpecific bool) (errs error) {
 	for _, hook := range hooks {
 		if !filepath.IsAbs(hook.Path) {
-			msgs = append(msgs, fmt.Sprintf("The %s hook %v: is not absolute path", hookType, hook.Path))
+			errs = multierror.Append(errs, fmt.Errorf("The %s hook %v: is not absolute path", hookType, hook.Path))
```

Comment:
The MUST for this check is [here][1].  For `poststart` and `poststop` the requirement is indirect, via [this][2] and [this][3].  Perhaps we should expand `Error.Reference` to an array of strings so we can represent trails like that?  Although if we're sticking to Markdown links, [this][4] is as specific as we can get at the moment, and it covers all the lines I linked earlier.

[1]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0/config.md#L375
[2]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0/config.md#L381
[3]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0/config.md#L383
[4]: https://github.com/opencontainers/runtime-spec/blob/v1.0.0/config.md#posix-platform-hooks

Reply:
I opened an issue to track this: https://github.com/opencontainers/runtime-tools/issues/452.